annotator-models/trainer/dawid_skene.py

```diff
@@ -0,0 +1,565 @@
+    items = list(df['_unit_id'].apply(lambda x: unit_id_to_index_map[x]))
+
+    y = list(df[label].apply(lambda x: y_to_index_map[x]))
+
+    nClasses = len(df[label].unique())
+    nItems = len(df['_unit_id'].unique())
+    nRaters = len(df['_worker_id'].unique())
+    counts = np.zeros([nItems, nRaters, nClasses])
+
+    # convert responses to counts
+    for i,item_index in enumerate(items):
+        rater_index = raters[i]
+        y_index = y[i]
+        counts[item_index,rater_index,y_index] += 1
+
+    unique_raters = index_to_worker_id_map.keys()
+    unique_items = index_to_unit_id_map.keys()
+    unique_classes = index_to_y_map.keys()
+
+    return unique_items, unique_raters, unique_classes, counts, index_to_unit_id_map
+
+def initialize(counts):
+    """
+    Get initial estimates for the true item classes using counts
+    see equation 3.1 in Dawid-Skene (1979)
+
+    Input:
+      counts: counts of the number of times each response was given
+          by each rater for each item: [items x raters x classes]. Note
+          in the crowd rating example, counts will be a 0/1 matrix.
+
+    Returns:
+      item_classes: matrix of estimates of true item classes:
+          [items x responses]
+    """
+    [nItems, nRaters, nClasses] = np.shape(counts)
+
+    # sum over raters
+    response_sums = np.sum(counts,1)
+
+    # create an empty array
+    item_classes = np.zeros([nItems, nClasses])
+
+    # for each item, take the average number of ratings in each class
+    for p in range(nItems):
+        item_classes[p,:] = response_sums[p,:] / np.sum(response_sums[p,:],dtype=float)
+
+    return item_classes
+
+def m_step(counts, item_classes):
+    """
+    Get estimates for the prior class probabilities (p_j) and the error
+    rates (pi_jkl) using MLE with current estimates of true item classes
+    See equations 2.3 and 2.4 in Dawid-Skene (1979)
+
+    Input:
+      counts: Array of how many times each rating was given by each rater
+        for each item
+      item_classes: Matrix of current assignments of items to classes
+
+    Returns:
+      p_j: class marginals [classes]
+      pi_kjl: error rates - the probability of rater k giving
+          response l for an item in class j [observers, classes, classes]
+    """
+    [nItems, nRaters, nClasses] = np.shape(counts)
+
+    # compute class marginals
+    class_marginals = np.sum(item_classes, 0)/float(nItems)
+
+    # compute error rates for each rater, each predicted class
+    # and each true class
+    error_rates = np.matmul(counts.T, item_classes)
+
+    # reorder axes so its of size [nItems x nClasses x nClasses]
+    error_rates = np.einsum('abc->bca', error_rates)
+
+    # divide each row by the sum of the error rates over all observation classes
+    sum_over_responses = np.sum(error_rates, axis=2)[:,:,None]
+    error_rates = np.divide(error_rates, sum_over_responses,
+                            where=sum_over_responses!=0)
```

Comment:
not sure we need to change this, but am concerned this may hide a bug (since it should never happen)

Reply:
I looked into this and `sum_over_responses` can have values that are 0 if a rater never recorded that

`sum_over_responses` is size `[nRaters, nClasses]`. If rater `k` never labeled any item with class `i`, then `sum_over_responses[k,i]` will be 0. In that case, we want the error rates to be 0, instead of being nan. 

```python
>>> sum_over_responses.shape
(588, 5, 1)

>>> sum_over_responses[587,:,:]
array([[0.5],
       [0.9],
       [0.6],
       [0. ],
       [0. ]])

# recall that error_rates[k, :, :] is a [nClasses, nClasses] array of the error rates for rater k
>>> error_rates[587,:,:]
array([[0.5, 0. , 0. , 0. , 0. ],
       [0.5, 0. , 0.4, 0. , 0. ],
       [0. , 0. , 0.6, 0. , 0. ],
       [0. , 0. , 0. , 0. , 0. ],
       [0. , 0. , 0. , 0. , 0. ]])

# divide error_rates by sum_over_responses, but not when the sum is 0
>>>  np.divide(error_rates, sum_over_responses, where=sum_over_responses!=0)[587,:,:]
array([[1.  , 0.  , 0.  , 0.  , 0.  ],
       [0.56, 0.  , 0.44, 0.  , 0.  ],
       [0.  , 0.  , 1.  , 0.  , 0.  ],
       [0.  , 0.  , 0.  , 0.  , 0.  ],
       [0.  , 0.  , 0.  , 0.  , 0.  ]])

# divide error_rates by sum_over_responses without the condition, see that we get nan
>>> np.divide(error_rates, sum_over_responses)[587,:,:]
array([[1.  , 0.  , 0.  , 0.  , 0.  ],
       [0.56, 0.  , 0.44, 0.  , 0.  ],
       [0.  , 0.  , 1.  , 0.  , 0.  ],
       [ nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan]])
```